felix/nftables/ipsets.go

```diff
@@ -0,0 +1,697 @@
+			continue
+		}
+
+		success = true
+		break
+	}
+	if !success {
+		s.logCxt.Panic("Failed to update IP sets after multiple retries.")
+	}
+}
+
+// tryResync attempts to bring our state into sync with the dataplane.  It scans the contents of the
+// IP sets in the dataplane and queues up updates to any IP sets that are out-of-sync.
+func (s *IPSets) tryResync() error {
+	// Log the time spent as we exit the function.
+	resyncStart := time.Now()
+	defer func() {
+		s.logCxt.WithFields(log.Fields{
+			"resyncDuration":           time.Since(resyncStart),
+			"ipSetsWithDirtyMembers":   s.ipSetsWithDirtyMembers.Len(),
+			"ipSetsToCreateOrRecreate": s.setNameToProgrammedMetadata.PendingUpdates().Len(),
+			"ipSetsToDelete":           s.setNameToProgrammedMetadata.PendingDeletions().Len(),
+		}).Debug("Finished IPSets resync")
+	}()
+
+	// Clear the dataplane metadata view, we'll build it back up again as we scan.
+	s.setNameToProgrammedMetadata.Dataplane().DeleteAll()
+
+	// Load sets from the dataplane. Update our Dataplane() maps with the actual contents
+	// of the data plane so that the next ApplyUpdates() call will be able to properly make
+	// incremental updates.
+	//
+	// For any set that doesn't match the desired data plane state, we'll queue up an update.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	sets, err := s.nft.List(ctx, "set")
+	if err != nil {
+		return fmt.Errorf("error listing nftables sets: %s", err)
+	}
+	for _, setName := range sets {
+		logCxt := s.logCxt.WithField("setName", setName)
+
+		elems, err := s.nft.ListElements(ctx, "set", setName)
+		if err != nil {
+			return fmt.Errorf("error listing nftables set elements: %s", err)
+		}
+		strElems := []string{}
+		for _, e := range elems {
+			if len(e.Key) == 3 {
+				// This is a concatination of IP, protocol and port. Format it back into Felix's internal representation.
+				strElems = append(strElems, fmt.Sprintf("%s,%s:%s", e.Key[0], e.Key[1], e.Key[2]))
+			} else {
+				// This is just an IP address / CIDR.
+				strElems = append(strElems, e.Key[0])
+			}
+		}
+
+		metadata, ok := s.setNameToAllMetadata[setName]
+		if !ok {
+			// Programmed in the data plane, but not in memory. Skip this one - we'll clean up
+			// state for this below.
+			continue
+		}
+		elemsSet := s.filterAndCanonicaliseMembers(metadata.Type, strElems)
+
+		memberTracker := s.getOrCreateMemberTracker(setName)
+		numExtrasExpected := memberTracker.PendingDeletions().Len()
+		err = memberTracker.Dataplane().ReplaceFromIter(func(f func(k ipsets.IPSetMember)) error {
+			elemsSet.Iter(func(item ipsets.IPSetMember) error {
+				f(item)
+				return nil
+			})
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("failed to read set memebers: %w", err)
+		}
+
+		// Mark us as having seen the programmed IP set.
+		// TODO: Ideally we'd extract this information from the data plane itself, but it's not exposed
+		// via knftables at the moment.
```

Comment:
hm, what does that mean?

Reply:
The in-memory metadata we keep for each set includes some things like the max size of the set. This function is attempting to synchronize our in-memory view of the set with what actually is programmed, but it's just copying the "desired" Size / Range metadata instead of reading it from the programmed set. I couldn't easily find a way to get this information from knft (can just get set names, and elements). 

I think it's OK and actually not especially relevant, considering right now the nftables code doesn't actually use any of these pieces of metadata when programming the set.

Reply:
Does nftables limit the size of sets by default? There's an option to set a limit, but perhaps the default is "no limit", which would be good for our use case.  It's an annoying footgun that iptables ipsets have a hard limit on size.

Reply:
To my knowledge, no it doesn't enforce a default limit.